src/test/python/complete_winml_tutorial.py

```diff
@@ -0,0 +1,149 @@
+import find_projection
+
+import pyrt
+import asyncio
+
+def timed_op(fun):
+    import time
+
+    def sync_wrapper(*args, **kwds):
+        print("Starting", fun.__name__)
+        start = time.perf_counter()
+
+        ret = fun(*args, **kwds)
+
+        end = time.perf_counter()
+        print(fun.__name__, "took", end - start, "seconds")
+        return ret
+
+    async def async_wrapper(*args, **kwds):
+        print("Starting", fun.__name__)
+        start = time.perf_counter()
+
+        ret = await fun(*args, **kwds)
+
+        end = time.perf_counter()
+        print(fun.__name__, "took", end - start, "seconds")
+        return ret
+
+    return async_wrapper if asyncio.iscoroutinefunction(fun) else sync_wrapper
+
+async def wrap_async_op(op):
+    loop = asyncio.get_event_loop()
+    future = loop.create_future()
+
+    def callback(operation, status):
+        if status == 1:
+            result = operation.GetResults()
+            loop.call_soon_threadsafe(asyncio.Future.set_result, future, result)
+        elif status == 2:
+            loop.call_soon_threadsafe(asyncio.Future.set_exception, future, asyncio.CancelledError())
+        elif status == 3:
+            loop.call_soon_threadsafe(asyncio.Future.set_exception, future, RuntimeError("AsyncOp failed"))
+        else:
+            loop.call_soon_threadsafe(asyncio.Future.set_exception, future, RuntimeError("Unexpected AsyncStatus"))
+
+    op.put_Completed(callback)
+
+    return await future
+
+def run_async_code(code):
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(code())
+    loop.close()
+
+import pyrt.Windows.AI.MachineLearning as winml 
+
+@timed_op
+def load_model(model_path):
+    if type(model_path) != str:
+        model_path = str(model_path)
+    return winml.LearningModel.LoadFromFilePath(model_path)
+
+@timed_op
+async def load_image_file(file_path):
+    if type(file_path) != str:
+        file_path = str(file_path)
```

Comment:
The best way to do this pattern is `os.fspath(file_path)`, which was specifically added to convert pathlib objects to str/bytes, but an unconditional `str(file_path)` is also fine.

Being able to do this implicitly would be nice too - most Python file APIs run their parameters through `fspath` (either directly or using the private C version).

Reply:
I'll look at changing the sample code to use os.fspath. But I'm not sure that we be implicitly converting arbitrary python objects to strings at the WinRT layer - i.e. auto converting model_path to a str *in the projection*.

Reply:
I'd consider `pathlib.*Path` objects for the same reason you'd do it for `datetime` - it's in the core library. There's a safe protocol for it though, so it probably fits neatly into the "when it's not a str, before raising see if it has `__fspath__` that returns str before raising". (Doing the same thing with `__str__` is not safe, since that's on every object.)

Reply:
For a bit more background, `pathlib`'s classes were _nearly_ subclasses of `str`. Some people's preference to use `bytes` for paths, as well as the difficulty in making existing operators return the correct type, prevented it. Plus a whole lot of people just thought it was dirty (not me :) )

Reply:
Done. THanks @zooba